124: Add localisation support

Polymehr opened the pull request:
#### Description

This fixes issue #124 
I added localisation support by surrounding most translatable strings with a `gettext` wrapper.

While refactoring the code I decided to make the following changes to it:
- I removed the `f`-strings and replaced them by `format()`-calls to avoid arbitrary code execution
- I used and added named braces to make reordering inserted terms and understanding them easier
- I removed the method `_cardS()` in `ngen.py` in favor of `ngettext()`
- I added `(s)` to a position where there may be multiple cards
- I kept some of the already existing translations in `main.py` and removed/changed the ones I thought of as unreasonable (e.g. the translation of the tool tip)
- I reformatted my inserted code to abide by Anki's maximal line length
- I used the old values as data for the `QtComboBox` when editing a card to avoid wrong behavior for some translations

To manage the translations I added the `lang.py` file.
Because I didn't know whether the threading support of Anki is important I adapted most code from Anki's `lang.py` file and tried abide by the restrictions of the license.
(Currently the add-on uses the locale Anki uses and this is not changeable.)

I added an empty `gettext`-`po` file for the German language at the correct location. The file was generated using `xgettext`. (I may contribute a German translation later.)
Because of the limitations of `aab` I placed the `locale` directory inside `src/image-occlusion-enhanced`.

Note: Because Python's `gettext`-API can only read compiled `po`-files (`mo` files) it may be a good next step to add automatic compilation of these files to `aab`.

Here is an example translation draft of the main UI (not pushed, download the `.mo` file [here](http://www.mediafire.com/file/h2arel3yv7s0o33/anki-image-occlusion-enhanced.mo/file)):

![translation example](https://user-images.githubusercontent.com/2074494/70842571-b35c1100-1e25-11ea-9078-1d6f07ab3add.png)

#### Checklist:

- [x] I've read and understood the [contribution guidelines](./CONTRIBUTING.md)
- [x] I've tested my changes against at least one of the following [Anki builds](https://apps.ankiweb.net/#download):
  - [x] Latest standard Anki 2.1 binary build [required for Anki-compatible 2.1 add-ons]
  - [x] Latest alternative Anki 2.1 binary build
  - [ ] Latest Anki 2.0 binary build [required for Anki 2.0-compatible add-ons]
- [x] I've tested my changes on at least one of the following platforms:
  - [x] Linux, version: Arch Linux (Anki 2.1.12 (eef86bf3))
  - [x] Windows, version: 10 (Anki 2.1.15 (442df9d6))
  - [ ] macOS, version: 
- [ ] My changes potentially affect non-desktop platforms, of which I've tested:
  - [ ] AnkiMobile, version:
  - [ ] AnkiDroid, version:
  - [ ] AnkiWeb

glutanimate commented:
Thanks for the exemplary PR! This looks very promising. However, with the add on being in a transitional and somewhat buggy state, I don’t feel comfortable merging a broad charge like this just yet. I need to take a closer look at this, so let me get back to you over the holiday break. (Generally speaking though, with any remaining concerns out of the way, I would be 100% on board with merging this).

Polymehr commented:
I made a few changes in the latest commit:
* I added complete German translation (with some help of @5hir0kur0), replacing the old file
* I moved `src/image_occlusion_enhanced/locale/de/LC_MESSAGES/anki-image-occlusion-enhanced.po` 
  to   `src/image_occlusion_enhanced/locale/anki-image-occlusion-enhanced.pot` thus also changing the extension from `.po` to `.pot` to properly mark it as a template file
* I found a fix a missing ‘>’ in 'dialogs.py' (“\<b>Fixing a broken note type:\</b\n”) I found while translating
* I also changed relative file paths in comments of `.po`(`t`) files, because they changed when the `.pot` file was moved

glutanimate commented:
Thanks, guys – both for this and the `aab` PR. All of this looks fantastic, but I definitely want to familiarize myself a bit more with Qt's translation system before jumping on board. I likely won't have time to do so for the next 2-3 weeks or so, but I'll try to follow up as soon as I can afterwards.

Polymehr commented:
I don't know whether I understand you correctly.

I only use Python's [`gettext`-package](https://docs.python.org/3/library/gettext.html).
Qt's translation system is not used. Anki also does not seem to use it directly and only packages the files required by the Qt-library. But that is just a guess by me.

In my opinion we shouldn't combine multiple systems because it is confusing.

To assist a little bit with your research: you maybe you want also want to have a look at the  [PO file format](https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html).

Polymehr commented:
I made a new optional commit that removes even more lines of Anki code from `lang.py` that aren't used by IOE currently.
To be more specific I removed all usages of multi threading because it isn't used by the add-on.

glutanimate commented:
Thanks again for this fantastic contribution! It seems like a few things were broken by some of the recent changes to Anki's i18n implementation – likely just a few minor adjustments needed to get things to a working state again, though.

I'm merging this into the `develop` branch now as I'm planning on reformatting the codebase with `black` and want to avoid causing too many merge conflicts.

Do you have time for / are you interested in adjusting the changes for more recent Anki releases (while maintaining compatibility with Anki 2.1.15 and up)? If so, I'd be more than happy to merge any follow-up PRs once `black` has run through.

Though if you don't have time that's absolutely ok as well of course. You've done more than enough already! I'd look into fixing things myself then. Either way is fine by me :)

glutanimate commented on src/image_occlusion_enhanced/lang.py:
```diff
@@ -0,0 +1,127 @@
+
+
+if not currentTranslation:
+    setLang(anki.lang.getLang())
```
Seems like `getLang` is no longer available in recent Anki releases. Not sure yet if there are any other incompatibilities.

5hir0kur0 commented:
For me `anki.lang.currentLang` is `None` for some reason, which is strange because it seems like it's set before the add-ons are loaded (in `setupLangAndBackend`).
```py
# qt/aqt/__init__.py
# i18n & backend
backend = setupLangAndBackend(pm, app, opts.lang)

# [...]

# load the main window
import aqt.main

mw = aqt.main.AnkiQt(app, pm, backend, opts, args)  # the add-ons are loaded here, right?
```

Anyway, if I change it to the following it seems to work for me:
```diff
diff --git i/src/image_occlusion_enhanced/lang.py w/src/image_occlusion_enhanced/lang.py
index 654905f..3c40ec5 100644
--- i/src/image_occlusion_enhanced/lang.py
+++ w/src/image_occlusion_enhanced/lang.py
@@ -30,9 +30,9 @@ Handle translation.
 
 import os
 import gettext
-import threading
 
 import anki.lang
+from aqt import mw
 
 # TODO maybe remove the table and 'mungeCode' once Anki 2.1.16 is released and
 #  replace with calls to the API to reduce copied code
@@ -124,5 +124,5 @@ def mungeCode(code):
 
 
 if not currentTranslation:
-    # FIXME:
-    setLang(anki.lang.currentLang)
+    # FIXME: Why is currentLang None?
+    setLang(anki.lang.currentLang or mw.pm.meta["defaultLang"] or "en_US")
```

Polymehr commented:
I will give it a try.
Thanks.

5hir0kur0 commented:
> Do you have time for / are you interested in adjusting the changes for more recent Anki releases (while maintaining compatibility with Anki 2.1.15 and up)? If so, I'd be more than happy to merge any follow-up PRs once `black` has run through.

I think I don't have time for this at the moment.  Maybe in a month or so but I can't promise it right now. I *am* interested, though :-)

Polymehr commented:
@5hir0kur0 
I just tested you proposal.
It seems to work for me.
`anki.lang.currentLang` _not_ is `None`for me but just `en`?

I tested that with Anki 2.1.26...

5hir0kur0 commented:
@Polymehr 
My language was set to German. I used Anki version 2.1.26 (downloaded from the website)
> Version 2.1.26 (70784154)
Python 3.8.1 Qt 5.14.1 PyQt 5.14.1

I also thought it's weird that it's `None`. Maybe it's a bug.
